File: src/main/java/com/hoi4utils/clausewitz/code/effect/Effect.java
package com.hoi4utils.clausewitz.code.effect;

import com.hoi4utils.clausewitz.code.scope.Scope;
import com.hoi4utils.clausewitz.code.scope.ScopeType;
import com.hoi4utils.clausewitz.script.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;


import java.util.EnumSet;
import java.util.function.Supplier;

/**
 * For information: <a href="https://hoi4.paradoxwikis.com/Effect">Effects
 * Wiki</a>
 */
public class Effect<T> extends DynamicPDX<T, StructuredPDX> implements ScopedPDXScript {

//    protected static final SortedMap<String, Effect<?>> effects = new TreeMap<>();
//
    public String identifier;
    private Scope defintionScope = null;
    private Scope targetScope;
    @NotNull private EnumSet<ScopeType> supportedScopes;
    @Nullable private EnumSet<ScopeType> supportedTargets;

    public Effect(String identifier, Supplier<PDXScript<T>> TSupplier, StructuredPDX structuredBlock) {
        super(TSupplier, structuredBlock);
        this.identifier = identifier;
    }

    /**
     * For simple effects that have no block version.
     * @param TSupplier
     */
    public Effect(String identifier, Supplier<PDXScript<T>> TSupplier) {
        super(TSupplier);
        this.identifier = identifier;
    }

    /**
     * For effects that only have a block definition version.
     * @param structuredBlock
     */
    public Effect(String identifier, StructuredPDX structuredBlock) {
        super(structuredBlock);
        this.identifier = identifier;
    }

    public String identifier() {
        return identifier;
    }

    public EnumSet<ScopeType> supportedScopes() {
        return supportedScopes;
    }

    public EnumSet<ScopeType> supportedTargets() {
		return supportedTargets;
	}

    public boolean hasSupportedTargets() {
		return (supportedTargets != null && !supportedTargets.isEmpty());
	}

	public void setTarget(Scope target) {
		this.targetScope = target;
	}

	public void setTarget(String string, Scope within) throws Exception {
		setTarget(Scope.of(string, within));
	}

    public Scope defintionScope() {
		return defintionScope;
	}

    public Scope targetScope() {
        return targetScope;
    }

    public String target() {
        if (targetScope == null) {
            return "[null target]";
        }
        return targetScope.name;
    }

    public boolean hasTarget() {
        return targetScope != null;
    }

    @Override
    public Scope getDefinitionScope() {
        return this.defintionScope;
    }

    public EnumSet<ScopeType> getSupportedScopes() {
        // Assuming there's a way to determine supported scopes
        return this.supportedScopes;
    }
//
//    @Override protected Object clone() throws CloneNotSupportedException {
//        Effect<?> c = (Effect<?>) super.clone();
//        c.parameters = new ArrayList<>();
//        return c;
//    }
}

File: src/main/scala/com/hoi4utils/clausewitz/code/effect/Effect.scala
package com.hoi4utils.clausewitz.code.effect

import com.hoi4utils.clausewitz.code.scope.{Scope, ScopeType}
import com.hoi4utils.clausewitz.script._
import org.jetbrains.annotations.NotNull

import scala.collection.mutable

// todo should be sealed? 
// todo  with AbstractPDX[Effect](identifiers) ????
trait Effect extends ScopedPDXScript with PDXScript[?] with Cloneable {
  
//  protected var tSupplier: () => ReferencePDX[Effect]
//  protected var structuredBlock: StructuredPDX

  protected var _definitionScope: Option[Scope] = None
  protected var _targetScope: Option[Scope] = None
  protected var _supportedScopes: Set[ScopeType] = Set.empty
  protected var _supportedTargets: Set[ScopeType] = Set.empty
  
  override def supportedScopes: Set[ScopeType] = _supportedScopes

  def supportedTargets: Set[ScopeType] = _supportedTargets

  def hasSupportedTargets: Boolean = _supportedTargets.nonEmpty

  def setTarget(target: Scope): Unit = {
    this._targetScope = Some(target)
  }

  def setTarget(string: String, within: Scope): Unit = {
    setTarget(Scope.of(string, within))
  }

  override def definitionScope: Option[Scope] = _definitionScope

  def targetScope: Option[Scope] = _targetScope

  def target: String = _targetScope.map(_.name).getOrElse("[null target]")

  def hasTarget: Boolean = _targetScope.isDefined

  def isScope: Boolean = false

  @throws[CloneNotSupportedException]
  override def clone(): AnyRef = {
    val clone = super.clone().asInstanceOf[Effect]
    clone._definitionScope = _definitionScope
    clone._targetScope = _targetScope
    clone._supportedScopes = _supportedScopes
    clone._supportedTargets = _supportedTargets
    clone
  }
}
